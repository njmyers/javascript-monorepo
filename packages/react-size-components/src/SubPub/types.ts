export type SubPub = {
  subscribe: () => null | { unsubscribe: () => boolean };
  destroy: () => any;
};

export type Options = {
  throttle?: boolean;
  debounce?: boolean;
  timeout?: number;
};
