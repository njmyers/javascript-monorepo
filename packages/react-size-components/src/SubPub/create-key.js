/** @flow */
const createKey = () =>
  Math.random()
    .toString(16)
    .slice(-8);

export default createKey;
