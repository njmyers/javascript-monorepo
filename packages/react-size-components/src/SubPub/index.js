/** @flow */
export { default } from './sub-pub'
export { default as isFunction } from './is-function'
export { default as createKey } from './create-key'
