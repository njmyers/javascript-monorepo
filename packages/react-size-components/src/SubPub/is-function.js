/** @flow */
const isFunction = (arg: any): boolean %checks => typeof arg === 'function';

export default isFunction;
