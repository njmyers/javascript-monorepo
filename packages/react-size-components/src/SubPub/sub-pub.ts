// @ts-nocheck
import throttle from 'lodash.throttle';
import debounce from 'lodash.debounce';
import fastdom from 'fastdom';

import isFunction from './is-function';
import createKey from './create-key';
import { SIXTY_FPS } from '../constants';
// types
import { SubPub as SB, Options } from './types';

const defaults: Options = {
  throttle: false,
  debounce: false,
  timeout: SIXTY_FPS,
};

const timingOptions = {
  leading: true,
  trailing: true,
};

function SubPub(eventType: string, _options: Options): SB {
  const subscribers = new Map();

  const options = {
    ...defaults,
    ...(_options || {}),
  };

  const publish = () => {
    subscribers.forEach(subscriber => fastdom.measure(() => subscriber()));
  };

  window.addEventListener(
    eventType,
    options.debounce
      ? debounce(publish, options.timeout, timingOptions)
      : options.throttle
      ? throttle(publish, options.timeout, timingOptions)
      : publish,
    { passive: true }
  );

  return Object.seal({
    subscribe(fn) {
      const key = createKey();

      if (isFunction(fn)) {
        subscribers.set(key, fn);
      } else {
        console.warn('tried to subscribe something that is not a function');
      }

      return {
        unsubscribe: () => subscribers.delete(key),
        update: fn => {
          if (isFunction(fn)) {
            subscribers.set(key);

            return true;
          } else {
            console.warn('tried to subscribe something that is not a function');

            return false;
          }
        },
      };
    },
    destroy() {
      window.removeEventListener(eventType, publish);
    },
  });
}

export default SubPub;
