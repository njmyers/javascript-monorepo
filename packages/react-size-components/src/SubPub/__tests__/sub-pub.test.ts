import SubPub from '../sub-pub';

describe('it works', () => {
  test('it is a function ', () => {
    expect(typeof SubPub).toBe('function');
  });

  test('it runs without crashing', () => {
    expect(SubPub()).toMatchSnapshot();
  });
});
