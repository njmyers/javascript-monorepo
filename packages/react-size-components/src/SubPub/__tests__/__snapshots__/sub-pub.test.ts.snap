// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it works it runs without crashing 1`] = `
Object {
  "destroy": [Function],
  "subscribe": [Function],
}
`;
